### PR TITLE
CLI Alonzo script support

### PR DESCRIPTION
**Pull request by dcoutts**

### CODE REVIEW

**Comment by dcoutts** [comment]
The current state here is that we cannot _yet_ construct valid Alonzo era txs. The next bit to make that work is to be able to provide the current protocol parameters to the `cardano-cli transaction build-raw` command.

In the CLI code we have these TODOs:
```
      TxBodyContent
        <$> validateTxIns  era inputsAndScripts
        <*> validateTxInsCollateral
                           era inputsCollateral
        <*> validateTxOuts era txouts
        <*> validateTxFee  era mFee
        <*> ((,) <$> validateTxValidityLowerBound era mLowerBound
                 <*> validateTxValidityUpperBound era mUpperBound)
        <*> validateTxMetadataInEra  era metadataSchema metadataFiles
        <*> validateTxAuxScripts     era scriptFiles
        <*> pure TxAuxScriptDataNone     --TODO alonzo: support this
        <*> pure TxExtraKeyWitnessesNone --TODO alonzo: support this
        <*> pure (BuildTxWith Nothing) --TODO alonzo: support this
        <*> validateTxWithdrawals    era withdrawals
        <*> validateTxCertificates   era certFiles
        <*> validateTxUpdateProposal era mUpdatePropFile
        <*> validateTxMintValue      era mValue
```
It's the 3rd one there that's critical. That's the protocol params. The others are non-critical features.

**Comment by dcoutts** [comment]
Ok, with any luck that'll now work. Needs to be tested.

**Review by Jimbo4350 — COMMENTED** [review]
I haven't tested this as yet but it LGTM

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs [hunk 1]
```diff
@@ -586,20 +599,70 @@ createScriptWitness
...
+        left $ ShelleyTxCmdScriptExpectedPlutus
+                 scriptFile
+                 (AnyScriptLanguage lang)
+
+
+readScriptDatumOrFile :: ScriptDatumOrFile witctx
+                      -> ExceptT ShelleyTxCmdError IO (ScriptDatum witctx)
+readScriptDatumOrFile (ScriptDatumOrFileForTxIn df) = ScriptDatumForTxIn <$>
+                                                        readScriptDataOrFile df
+readScriptDatumOrFile NoScriptDatumOrFileForMint    = pure NoScriptDatumForMint
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:695]: The consistent use of the witctx type argument should ensure that we're using only the right one in the right context. So it should be impossible to supply a datum for minting or stake contexts. For starters, there's no cli flag for it: we only have the datum cli flag for the txin context.

[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:695]: I meant from the perspective of the ledger. The ledger doesn't care if we supply a datum when using a minting or stake script. IIRC Lars was using tx datums in minting scrips.

**Review by Jimbo4350 — APPROVED** [review]

**Comment by dcoutts** [comment]
bors merge

**Pull request merged**